lib/adapters/tool/playwright/index.ts

```diff
@@ -0,0 +1,311 @@
+        return this._htmlReporter;
+    }
+
+    get guiApi(): GuiApi | undefined {
+        return this._guiApi;
+    }
+
+    initGuiApi(): void {
+        this._guiApi = GuiApi.create();
+    }
+
+    async readTests(): Promise<PlaywrightTestCollectionAdapter> {
+        const stdout = await new Promise<string>((resolve, reject) => {
+            // specify default browser in order to get correct stdout with browser name
+            const browserArgs = this._haveProjectsInConfig ? [] : ['--browser', DEFAULT_BROWSER_ID];
+
+            const child = spawn(this._pwtBinaryPath, ['test', '--list', '--reporter', 'list', '--config', this._configPath, ...browserArgs]);
+            let stdout = '';
+            let stderr = '';
+
+            child.stdout.on('data', (data) => stdout += data);
+            child.stderr.on('data', (data) => stderr += data);
+
+            child.on('error', (error) => reject(error));
+
+            child.on('exit', (code) => {
+                if (code !== 0) {
+                    return reject(new Error(`Playwright process with reading tests exited with code: ${code}, stderr: ${stderr}`));
+                }
+
+                resolve(stdout);
+            });
+        });
+
+        const stdoutByLine = stdout.split('\n').map(v => v.trim());
+        const startIndex = stdoutByLine.findIndex(v => v === 'Listing tests:');
+        const endIndex = stdoutByLine.findIndex(v => /total: \d+ tests? in \d+ file/i.test(v));
+
+        const tests = stdoutByLine.slice(startIndex + 1, endIndex).map(line => {
+            const [browserName, file, ...titlePath] = line.split(PWT_TITLE_DELIMITER);
+
+            return {
+                browserName: browserName.slice(1, -1),
+                file: file.split(':')[0],
+                title: titlePath.join(DEFAULT_TITLE_DELIMITER),
+                titlePath
+            } as PwtRawTest;
+        });
+
+        return PlaywrightTestCollectionAdapter.create(tests);
+    }
+
+    async run(_testCollection: PlaywrightTestCollectionAdapter, tests: TestSpec[] = []): Promise<boolean> {
+        return this._runTests(tests);
+    }
+
+    async runWithoutRetries(_testCollection: PlaywrightTestCollectionAdapter, tests: TestSpec[] = []): Promise<boolean> {
+        return this._runTests(tests, ['--retries', '0']);
+    }
+
+    private async _runTests(tests: TestSpec[] = [], runArgs: string[] = []): Promise<boolean> {
+        if (!this._reportBuilder || !this._eventSource) {
+            throw new Error('"reportBuilder" and "eventSource" instances must be initialize before run tests');
+        }
+
+        const queue = new PQueue({concurrency: os.cpus().length});
+
+        return new Promise((resolve, reject) => {
+            const args = ([] as string[]).concat(prepareRunArgs(tests, this._configPath, this._haveProjectsInConfig), runArgs);
+            const child = spawn(this._pwtBinaryPath, args, {
+                stdio: ['inherit', 'inherit', 'inherit', 'ipc']
```

Comment:
Use ipc in order to communicate between special reporter which is started inside pwt and html-reporter process.